### PR TITLE
Implement file logging with default xunit tests

### PR DESCRIPTION
**Pull request by JunTaoLuo**
This is needed for https://github.com/aspnet/KestrelHttpServer/issues/2390:

The usage is:

1. Extend base class of `LoggedTest`
```c#
    public class MyTests : LoggedTest
    {
    }
```
2. Use the properties of the base class `LoggedTest`
```c#
        [Fact]
        public void MyTest()
        {
            // using LoggerFactory
            using (var webhost = WebHost.CreateDefaultBuilder().ConfigureServices(AddTestLogging).Build())
            {
                //...
            }
            // using Logger
            Logger.LogInformation("Hello world");
            // using TestOutputHelper
           new XunitLogger(TestOutpuHelper, nameof(MyClass), LogLevel.Information);
            // using TestSink
            TestSink.Writes.ForEach(c => Assert.True(c.LogLevel >= LogLeve.Debug));
        }
```
3. The verbosity can be controlled via the LogLevelAttribute:
```c#
        [Fact]
        [LogLevel(LogLevel.Information)]
        public void MyTest()
        {
            // This will noop
            Logger.Debug("Hello world");
        }
```

### CODE REVIEW

File: src/Microsoft.Extensions.Logging.Testing/Xunit/Shared/SkippedTestCase.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNetCore.Testing.xunit
```
Comment (JunTaoLuo) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/Shared/SkippedTestCase.cs]: Going to make this type public from testing instead of duplicating here.

File: test/Microsoft.Extensions.Logging.Testing.Tests/LoggedTestXunitTests.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        public LoggedTestXunitTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public void LoggedFactGetsInitializedLoggerFactory()
+        {
+            Assert.NotNull(LoggerFactory);
+            LoggerFactory.CreateLogger(nameof(AssemblyTestLogTests)).LogInformation("Hello world");
+            LoggerFactory.CreateLogger(nameof(AssemblyTestLogTests)).LogDebug("Debug Hello world");
```
Comment (JunTaoLuo) [inline test/Microsoft.Extensions.Logging.Testing.Tests/LoggedTestXunitTests.cs]: TODO: add more tests

File: src/Microsoft.Extensions.Logging.Testing/Microsoft.Extensions.Logging.Testing.csproj [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
   <PropertyGroup>
     <Description>Helpers for writing tests that use Microsoft.Extensions.Logging. Contains null implementations of the abstractions that do nothing, as well as test implementations that are observable.</Description>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
```
Comment (JunTaoLuo) [inline src/Microsoft.Extensions.Logging.Testing/Microsoft.Extensions.Logging.Testing.csproj]: I needed to cross-target to enable test discovery on full framework.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/Shared/SkippedTestCase.cs (hunk 1 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/Shared/SkippedTestCase.cs]: https://github.com/aspnet/Testing/pull/301

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs [hunk 4]
```diff
@@ -20,6 +22,15 @@ public class AssemblyTestLog : IDisposable
...
         private static readonly int MaxPathLength = 245;
+        private static IEnumerable<char> InvalidFileChars =
+            Path.GetInvalidPathChars()
+            .Union(Path.GetInvalidFileNameChars())
+            .Union(new char[] {
+                ' ', // space
+                '\\', // back slash
+                '/', // forward slash
+                '\x7F', // delete
+            });
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: ToArray

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs [hunk 5]
```diff
@@ -234,6 +253,26 @@ private static SerilogLoggerProvider ConfigureFileLogging(string fileName)
...
 
+        private static string RemoveIllegalFileChars(string s)
+        {
+            var sb = new StringBuilder();
+
+            foreach (var c in s)
+            {
+                if (InvalidFileChars.Contains(c))
+                {
+                    var escapedChar = Convert.ToByte(c).ToString("x");
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: This expands string size a lot, maybe just do '_' ?

Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: I did do that at one point but it made test names highly unreadable. Also converting to the same character for all invalid chars creates lots of collisions. With the test name shortening logic in place, I don't think this should be an issue.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestFrameworkDiscoverer.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+using Microsoft.AspNetCore.Testing.xunit;
+using Xunit;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Microsoft.Extensions.Logging.Testing
+{
+    public class LoggedTestFrameworkDiscoverer : XunitTestFrameworkDiscoverer
+    {
+        private List<(Type Type, IXunitTestCaseDiscoverer Discoverer)> Discoverers { get; }
```
Comment (anurse) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestFrameworkDiscoverer.cs]: `Dictionary`?

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestCaseRunner.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+using System.Reflection;
+using System.Threading;
+using Xunit.Abstractions;
+using Xunit.Sdk;
+
+namespace Microsoft.Extensions.Logging.Testing
+{
+    public class LoggedTestCaseRunner : XunitTestCaseRunner
+    {
+        public LoggedTestCaseRunner(IXunitTestCase testCase, string displayName, string skipReason, object[] constructorArguments, object[] testMethodArguments, IMessageBus messageBus, ExceptionAggregator aggregator, CancellationTokenSource cancellationTokenSource) : base(testCase, displayName, skipReason, constructorArguments, testMethodArguments, messageBus, aggregator, cancellationTokenSource)
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestCaseRunner.cs]: Linebreaks? This line spans 2 monitors )

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        public LoggedTestInvoker(ITest test, IMessageBus messageBus, Type testClass, object[] constructorArguments, MethodInfo testMethod, object[] testMethodArguments, IReadOnlyList<BeforeAfterTestAttribute> beforeAfterAttributes, ExceptionAggregator aggregator, CancellationTokenSource cancellationTokenSource) : base(test, messageBus, testClass, constructorArguments, testMethod, testMethodArguments, beforeAfterAttributes, aggregator, cancellationTokenSource)
+        {
+        }
+
+        protected override object CreateTestClass()
+        {
+            var testClass = base.CreateTestClass();
+
+            if (testClass is LoggedTest loggedTestClass)
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: Can't we do it for any class, so you're not forced to inherit and can just inject the constructor argument?

File: test/Microsoft.Extensions.Logging.Testing.Tests/LoggedTestXunitTests.cs [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+using Microsoft.AspNetCore.Testing.xunit;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.Extensions.Logging.Testing.Tests
+{
+    public class LoggedTestXunitTests : LoggedTest
+    {
+        public LoggedTestXunitTests(ITestOutputHelper output) : base(output)
```
Comment (pakrym) [inline test/Microsoft.Extensions.Logging.Testing.Tests/LoggedTestXunitTests.cs]: Do we still have to pass `ITestOutputHelper` around? Can ILoggerFactory be injected directly?

**Comment by JunTaoLuo** [comment]
Note: WIP adding and fixing tests.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs (hunk 5 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: Hmm right now I expand illegal characters like `/` => `-x2f`, maybe I can reduce the verbosity and just leave it as `2f` which is a bit harder to parse out but still retains most of the information.

Comment (pakrym) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: > I did do that at one point but it made test names highly unreadable

Hex is not very readable too.

> With the test name shortening logic in place, I don't think this should be an issue. 

Replacing a bunch of '_' with one would be a nice addition to shortening logic.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            {
+                var classType = loggedTestClass.GetType();
+                var testOutputHelper = ConstructorArguments.Single(a => typeof(ITestOutputHelper).IsAssignableFrom(a.GetType())) as ITestOutputHelper;
```
Comment (anurse) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: In theory we could get the `ITestOutputHelper` from the source (whatever that is) rather than extracting it from constructor arguments.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+                var logLevelAttribute = TestMethod.GetCustomAttribute(typeof(LogLevelAttribute)) as LogLevelAttribute;
+                var testName = TestMethodArguments.Aggregate(TestMethod.Name, (a, b) => $"{a}_{(b ?? "null")}");
```
Comment (anurse) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: `string.Join('_', ...)`?

**Comment by pakrym** [comment]
Supporting https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Extensions.Logging.Testing/ITestSink.cs would be interesting too, would make some kestrel tests even simpler

**Comment by anurse** [comment]
Ooh, yeah, automatically register a TestLogger and TestSink. I like that. If it's simple, do it, if it's more than a couple minutes work, file a bug and move on :)

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestFrameworkDiscoverer.cs (hunk 6 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestFrameworkDiscoverer.cs]: Ah yes, at one point, I was using a best match on the type which needed a list of discoverers for types from the most specific to the most general. Now that I'm using an exact match on the type and falling back to non-logged discoverers, this can totally be just a dictionary.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs (hunk 5 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: Poor choice of words, I meant file name collisions instead of readability. Even though we have collision resolution, I would prefer retaining as much information as possible regarding the arguments used by default.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs (hunk 8 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: We are not injecting it as a constructor argument. We are setting the property of the base class LoggedTest.

Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: Unless you are suggesting we change the design?

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs (hunk 5 above)
Comment (pakrym) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: Do you really prefer `Testing_https:-x2f-x2flocalhost` over `Testing_https:_localhost` ? To me, the second one is much easier to figure out.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs (hunk 8 above)
Comment (pakrym) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: I always thought we are going argument injection, that's why we started with argument injection DI prototype.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs (hunk 5 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: For that case maybe, but this will create many collisions for kestrel tests, for example: https://github.com/aspnet/KestrelHttpServer/blob/dev/test/shared/HttpParsingData.cs#L192. Also I don't like coalescing multiple `_` since that makes tests like https://github.com/aspnet/KestrelHttpServer/blob/dev/test/shared/HttpParsingData.cs#L254-L255 indistiguishable.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs (hunk 11 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: Closest I can come up with is `string.Join("_", TestMethod.Name, TestMethodArguments)` but that won't be able to mark `null` object with `"null"` string

Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: From adding tests previously, I remember it being useful to distinguish between null strings and empty strings for example. That's why I special cased null here.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs [hunk 12]
```diff
@@ -20,6 +22,15 @@ public class AssemblyTestLog : IDisposable
         public static readonly string OutputDirectoryEnvironmentVariableName = "ASPNETCORE_TEST_LOG_DIR";
         private static readonly string LogFileExtension = ".log";
         private static readonly int MaxPathLength = 245;
+        private static char[] InvalidFileChars =
+            Path.GetInvalidPathChars()
+            .Union(Path.GetInvalidFileNameChars())
+            .Union(new char[] {
+                ' ', // space
+                '\\', // back slash
+                '/', // forward slash
```
Comment (halter73) [inline src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: Isn't this already an InvalidFileNameChar on every platform? Maybe a whitelist would be better.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+            : base(diagnosticMessageSink)
+        {
+        }
+
+        protected override IEnumerable<IXunitTestCase> CreateTestCasesForTheory(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo theoryAttribute)
+        {
+            var skipReason = testMethod.EvaluateSkipConditions();
+            return skipReason != null
+               ? new[] { new SkippedTestCase(skipReason, DiagnosticMessageSink, discoveryOptions.MethodDisplayOrDefault(), testMethod) }
+               : base.CreateTestCasesForTheory(discoveryOptions, testMethod, theoryAttribute);
```
Comment (halter73) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs]: Does the base class not call EvaluateSkipConditions?

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs (hunk 11 above)
Comment (halter73) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: 👍 for foldl

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs (hunk 12 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: I think so. I prefer not using a whitelist since that probably takes more work.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs (hunk 11 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: What is foldl?

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs (hunk 13 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs]: Not LoggedTheoryDiscoverer.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs (hunk 5 above)
Comment (pakrym) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: @halter73 what would you prefer?

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs (hunk 13 above)
Comment (halter73) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs]: But isn't that what you're replacing? If this didn't need to call EvaluateSkipConditions before, why now?

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs (hunk 11 above)
Comment (halter73) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: The hipster name for Aggregate.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs (hunk 13 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs]: No I'm replacing ConditionalTheoryDiscoverer. But this inherits from LoggedTheoryDiscoverer which doesn't call EvaluateSkipConditions.

Comment (halter73) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs]: Wouldn't ConditionalTheoryDiscoverer have it's own LoggedTheoryDiscoverer that does exactly this then? Why not use that?

File: test/Microsoft.Extensions.Logging.Testing.Tests/LoggedTestXunitTests.cs (hunk 9 above)
Comment (JunTaoLuo) [reply test/Microsoft.Extensions.Logging.Testing.Tests/LoggedTestXunitTests.cs]: Actually I don't think we can remove ITestOutputHelper. What if the user wants to initialize a different LoggerFactory by calling StartLog (to override test name for example)? That call needs the ITestOutputHelper.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs (hunk 13 above)
Comment (halter73) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs]: I think I misunderstood. Could you derive from ConditionalTheoryDiscoverer?

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs (hunk 8 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: Found a way of removing the need to flow the ITestOutputHelper with constructors.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs (hunk 13 above)
Comment (JunTaoLuo) [reply src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedConditionalTheoryDiscoverer.cs]: Then I'll still need to override these two methods to create `LoggedTestCase` or `LoggedTheoryTestCase`. I don't see the benefit here.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs [hunk 14]
```diff
@@ -20,6 +22,14 @@ public class AssemblyTestLog : IDisposable
         public static readonly string OutputDirectoryEnvironmentVariableName = "ASPNETCORE_TEST_LOG_DIR";
         private static readonly string LogFileExtension = ".log";
         private static readonly int MaxPathLength = 245;
+        private static char[] InvalidFileChars = new char[]
+        {
+            '\"', '<', '>', '|', '\0',
```
Comment (JunTaoLuo) [inline src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: Created a static blacklist by concatenating https://github.com/dotnet/coreclr/blob/f5ae6d14d9d49dbb9f4e97f605d2a5cceab04879/src/mscorlib/shared/System/IO/Path.Unix.cs#L13-L15 and https://github.com/dotnet/coreclr/blob/9ac41819ab05282798c498b373a5017b1c2c4e64/src/mscorlib/shared/System/IO/Path.Windows.cs#L12-L28 
as well as space and delete.

**Comment by pakrym** [comment]
Can we bring `[assembly: Xunit.TestFramework("Microsoft.Extensions.Logging.Testing.LoggedTestFramework", "Microsoft.Extensions.Logging.Testing")]` via MSBuild target?

**Comment by pakrym** [comment]
We can also add a method to `LoggetTest` to change
```
 [Fact]
        public void MyTest()
        {
            // using LoggerFactory
            using (var webhost = WebHost.CreateDefaultBuilder().ConfigureServices(collection => collection.AddSingleton(LoggerFactory)).Build())
            {
                //...
            }

```
to
```
[Fact]
        public void MyTest()
        {
            // using LoggerFactory
            using (var webhost = WebHost.CreateDefaultBuilder().ConfigureServices(AddTestLogging).Build())
            {
                //...
            }
```
Makes it a bit shorter and simpler.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs [hunk 15]
```diff
@@ -234,6 +252,17 @@ private static SerilogLoggerProvider ConfigureFileLogging(string fileName)
             return new SerilogLoggerProvider(serilogger, dispose: true);
         }
 
+        private static string RemoveIllegalFileChars(string s)
+        {
+            var sb = new StringBuilder();
+
+            foreach (var c in s)
+            {
+                sb.Append(InvalidFileChars.Contains(c) ? Convert.ToByte(c).ToString("x") : $"{c}");
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: Not when we have space in the invalid character list, please, let's not use hex, a bunch of random lower case characters would make file names completely unreadable.

File: src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+
+                // Try resolving ITestOutputHelper from constructor arguments
+                loggedTestClass.TestOutputHelper = ConstructorArguments?.SingleOrDefault(a => typeof(ITestOutputHelper).IsAssignableFrom(a.GetType())) as ITestOutputHelper;
+
+                // None resolved so create a new one and retain a reference to it for initialization/uninitialization
+                if (loggedTestClass.TestOutputHelper == null)
+                {
+                    loggedTestClass.TestOutputHelper = _output = new TestOutputHelper();
```
Comment (pakrym) [inline src/Microsoft.Extensions.Logging.Testing/Xunit/LoggedTestInvoker.cs]: Don't you have to call
 https://github.com/xunit/xunit/blob/bccfcccf26b2c63c90573fe1a17e6572882ef39c/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestRunner.cs#L65
 for `TestOutputHelper` to work?

**Comment by pakrym** [comment]
Couple of comments then LGTM, let's try to use it.

File: src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs (hunk 15 above)
Comment (pakrym) [reply src/Microsoft.Extensions.Logging.Testing/AssemblyTestLog.cs]: Only append underscore if the previous character is not underscore. `A_B` is better than `A_____B` and might help long paths.

**Comment by JunTaoLuo** [comment]
Addressed the feedback items.

**Comment by anurse** [comment]
🎉
